pylint/checkers/modified_iterating_checker.py

```diff
@@ -0,0 +1,153 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+
+from typing import TYPE_CHECKING, Optional
+
+from astroid import nodes
+
+from pylint import checkers, interfaces
+from pylint.checkers import utils
+
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+
+_LIST_MODIFIER_METHODS = {"append", "remove"}
+_SET_MODIFIER_METHODS = {"add", "remove"}
+
+
+class ModifiedIterationChecker(checkers.BaseChecker):
+    """Checks for modified iterators in for loops iterations.
+
+    Currently supports `for` loops for Sets, Dictionaries and Lists.
+    """
+
+    __implements__ = interfaces.IAstroidChecker
+
+    name = "modified_iteration"
+
+    msgs = {
+        "W4701": (
+            "Iterated list '%s' is being modified inside for loop body, consider iterating through a copy of it "
+            "instead.",
+            "modified-iterating-list",
+            "Emitted when items are added or removed to a list being iterated through. "
+            "Doing so can result in unexpected behaviour, that is why it is preferred to use a copy of the list.",
+        ),
+        "E4702": (
+            "Iterated dict '%s' is being modified inside for loop body, iterate through a copy of it instead.",
+            "modified-iterating-dict",
+            "Emitted when items are added or removed to a dict being iterated through. "
+            "Doing so raises a RuntimeError.",
+        ),
+        "E4703": (
+            "Iterated set '%s' is being modified inside for loop body, iterate through a copy of it instead.",
+            "modified-iterating-set",
+            "Emitted when items are added or removed to a set being iterated through. "
+            "Doing so raises a RuntimeError.",
+        ),
+    }
+
+    options = ()
+    priority = -2
+
+    @utils.check_messages(
+        "modified-iterating-list", "modified-iterating-dict", "modified-iterating-set"
+    )
+    def visit_for(self, node: nodes.For) -> None:
+        iter_obj = node.iter
+        for body_node in node.body:
+            message_raised = self._modified_iterating_check(body_node, iter_obj)
+            if not message_raised:  # needs further check when message not raised
+                for child in body_node.get_children():
+                    self._modified_iterating_check(child, iter_obj)
+
+    def _modified_iterating_check(self, node, iter_obj):
```

Comment:
```suggestion
    def _modified_iterating_check(self, node: ?, iter_obj: ?) -> bool:
```
I don't know about the type of ``body_node`` maybe NodeNG ? for iter_obj there's probably a particular node class.

Reply:
In my simple that it can be all of: `ListComp`, `Name`, `Cal` .. so I will stick with `NodeNG`..